### PR TITLE
[DO-1892] Transliteration

### PR DESCRIPTION
**Pull request by Iva-khar**
For testing run `python manage.py test data_ocean.tests.TransliterateTestCase`

### CODE REVIEW

File: data_ocean/utils.py [hunk 1]
```diff
@@ -150,3 +153,195 @@ def replace_incorrect_symbols(string):
...
+    'ч': 'ch',
+    'ш': 'sh',
+    'щ': 'shch',
+    'ъ': '',
+    'ы': 'y',
+    'ь': '',
+    'э': 'e',
+    'ю': 'iu',
+    'я': 'ia',
+    'А': 'A',
```
Comment (criticus) [inline data_ocean/utils.py:273]: how is it different from the previous dictionary?

File: data_ocean/utils.py [hunk 2]
```diff
@@ -150,3 +153,195 @@ def replace_incorrect_symbols(string):
...
+    company_type = CompanyType.objects.filter(name=string[0].strip()).first().name_eng if \
+        CompanyType.objects.filter(name=string[0].strip()).first() else transliterate(string[0].strip())
+    return (company_type + ' ' + transliterate('"' + '"'.join(string[1:]))).strip()
+
+
+def transliterate_address(string):
+    string = string.split(',')
+    country = Country.objects.filter(name_uk=string[0].strip().lower()).first().name if \
+        Country.objects.filter(name_uk=string[0].strip().lower()).first() else transliterate(string[0].strip())
+    return (country + ', ' + transliterate(','.join(string[1:]).lower())).strip()
```
Comment (roman-tiukh) [inline data_ocean/utils.py:347]: I think it should be one universal function (except country finding in text). We must write a function for every field in the system?

File: data_ocean/utils.py [hunk 3]
```diff
@@ -150,3 +153,195 @@ def replace_incorrect_symbols(string):
...
+                new_string += trans_char(char, ukr_to_en_first_word)
+            else:
+                new_string += trans_char(char, ukr_to_en)
+    return new_string
+
+
+def transliterate_company_name(string):
+    string = string.split('"')
+    company_type = CompanyType.objects.filter(name=string[0].strip()).first().name_eng if \
+        CompanyType.objects.filter(name=string[0].strip()).first() else transliterate(string[0].strip())
```
Comment (roman-tiukh) [inline data_ocean/utils.py:339]: 2 requests for DB, we can only one...

File: data_ocean/utils.py [hunk 4]
```diff
@@ -150,3 +153,195 @@ def replace_incorrect_symbols(string):
...
+    return (company_type + ' ' + transliterate('"' + '"'.join(string[1:]))).strip()
+
+
+def transliterate_address(string):
+    string = string.split(',')
+    country = Country.objects.filter(name_uk=string[0].strip().lower()).first().name if \
+        Country.objects.filter(name_uk=string[0].strip().lower()).first() else transliterate(string[0].strip())
```
Comment (roman-tiukh) [inline data_ocean/utils.py:346]: same(

File: data_ocean/utils.py [hunk 5]
```diff
@@ -150,3 +153,195 @@ def replace_incorrect_symbols(string):
...
+def trans_char(char, transliterate_dict):
+    for key in transliterate_dict:
+        if key == char:
+            return transliterate_dict[key]
+        elif key == 'Є':
+            return char
+
+
+def transliterate(string):
+    string = string.strip(' ')
```
Comment (roman-tiukh) [inline data_ocean/utils.py:323]: strip()???

File: data_ocean/utils.py [hunk 6]
```diff
@@ -150,3 +153,195 @@ def replace_incorrect_symbols(string):
...
+    'Є': 'Ye'
+}
+
+
+def trans_char(char, transliterate_dict):
+    for key in transliterate_dict:
+        if key == char:
+            return transliterate_dict[key]
+        elif key == 'Є':
+            return char
```
Comment (roman-tiukh) [inline data_ocean/utils.py:319]: :neutral_face: 
```suggestion
    return transliterate_dict.get(char, char)
```

File: data_ocean/utils.py [hunk 7]
```diff
@@ -150,3 +153,195 @@ def replace_incorrect_symbols(string):
...
+            else:
+                new_string += trans_char(char, ukr_to_en)
+    return new_string
+
+
+def transliterate_company_name(string):
+    string = string.split('"')
+    company_type = CompanyType.objects.filter(name=string[0].strip()).first().name_eng if \
+        CompanyType.objects.filter(name=string[0].strip()).first() else transliterate(string[0].strip())
+    return (company_type + ' ' + transliterate('"' + '"'.join(string[1:]))).strip()
```
Comment (roman-tiukh) [inline data_ocean/utils.py:340]: what if address don't have country?

**Review by roman-tiukh — CHANGES_REQUESTED** [review]
For good result, you should use regexp

File: data_ocean/utils.py (hunk 2 above)
Comment (Iva-khar) [reply data_ocean/utils.py]: rewrited

File: data_ocean/utils.py (hunk 3 above)
Comment (Iva-khar) [reply data_ocean/utils.py]: done

File: data_ocean/utils.py (hunk 4 above)
Comment (Iva-khar) [reply data_ocean/utils.py]: done

File: data_ocean/utils.py (hunk 5 above)
Comment (Iva-khar) [reply data_ocean/utils.py]: fixed

File: data_ocean/utils.py (hunk 6 above)
Comment (Iva-khar) [reply data_ocean/utils.py]: implemented

File: data_ocean/utils.py (hunk 7 above)
Comment (Iva-khar) [reply data_ocean/utils.py]: just transliterate will

**Comment by Iva-khar** [comment]
Add testing

**Review by roman-tiukh — COMMENTED** [review]
just use SimpleTestCase

File: data_ocean/tests.py [hunk 8]
```diff
@@ -1,3 +1,39 @@
...
+        company_name = transliterate_field(self.testing_company.name, 'company_name')
+        address = transliterate_field(self.testing_company.address, 'address')
+        if re.search("[А-ЩЬЮЯҐЄІЇа-щьюяґєії]", company_name):
+            company_name = None
+        if re.search("[А-ЩЬЮЯҐЄІЇа-щьюяґєії]", address):
+            address = None
+        print(company_name)
+        print(address)
+        self.assertEqual(transliterate_field(self.testing_company.name, 'company_name'), company_name)
+        self.assertEqual(transliterate_field(self.testing_company.address, 'address'), address)
```
Comment (roman-tiukh) [inline data_ocean/tests.py:39]: completely incorrect test

**Review by roman-tiukh — CHANGES_REQUESTED** [review]

File: data_ocean/tests.py [hunk 9]
```diff
@@ -1,3 +1,39 @@
...
+        pass
+
+    def teardown_databases(self, old_config, **kwargs):
+        """ Override the database teardown defined in parent class """
+        pass
+
+
+class TransliterateTestCase(TestCase):
+    def setUp(self):
+        company_ids = list(Company.objects.values('id'))
```
Comment (criticus) [inline data_ocean/tests.py:25]: if we need just one company there is no point to get all and select one, consider using a constant value

**Comment by criticus** [comment]
what is the reason for writing our own functions instead of using https://pypi.org/project/transliterate/ ?

File: data_ocean/tests.py (hunk 9 above)
Comment (Iva-khar) [reply data_ocean/tests.py:25]: we gave up the use of db registers in tests

**Comment by Iva-khar** [comment]
> what is the reason for writing our own functions instead of using https://pypi.org/project/transliterate/ ?

This pack not handle with different transliteration same letters depend of position in word. In ukrainian 5 letters with different transliteration depends of it's first position in word or others

**Comment by criticus** [comment]
> > what is the reason for writing our own functions instead of using https://pypi.org/project/transliterate/ ?
> 
> This pack not handle with different transliteration same letters depend of position in word. In ukrainian 5 letters with different transliteration depends of it's first position in word or others

I believe you can use `pre_processor_mapping` to replace characters at the beginning of words

**Comment by criticus** [comment]
for example, see https://github.com/barseghyanartur/transliterate/issues/52

**Comment by roman-tiukh** [comment]
@criticus I will check this

File: data_ocean/utils.py [hunk 10]
```diff
@@ -150,3 +153,195 @@ def replace_incorrect_symbols(string):
         .replace("''", '"')
     string = re.sub(r'\s+', ' ', string)
     return string
+
+
+ukr_to_en_first_word = {
```
[the remark this answers is not in the document]
Comment (Iva-khar) [reply data_ocean/utils.py]: done

File: data_ocean/utils.py (hunk 1 above)
Comment (Iva-khar) [reply data_ocean/utils.py]: fist dictionary for the first letter in the word, second dictionary for oters words

File: data_ocean/tests.py (hunk 8 above)
Comment (Iva-khar) [reply data_ocean/tests.py]: rewritted

File: data_ocean/tests.py [hunk 11]
```diff
@@ -1,3 +1,97 @@
...
+                'Україна, 36014, Полтавська обл., місто Полтава, ПЛОЩА ПАВЛЕНКІВСЬКА, будинок 24'
+            ),
+            'Ukraine, 36014, Полтавська обл., місто Полтава, ПЛОЩА ПАВЛЕНКІВСЬКА, будинок 24'
+        )
+        self.assertEqual(
+            translate_country_in_string(
+                'Україна, 32500, Хмельницька обл., Віньковецький р-н, смт. Віньківці, ВУЛИЦЯ СОБОРНА УКРАЇНA'
+            ),
+            'Ukraine, 32500, Хмельницька обл., Віньковецький р-н, смт. Віньківці, ВУЛИЦЯ СОБОРНА УКРАЇНA'
+        )
```
Comment (roman-tiukh) [inline data_ocean/tests.py:20]: more pretty and readable
image

**Review by roman-tiukh — CHANGES_REQUESTED** [review]

File: data_converter/settings_local.base.py [hunk 12]
```diff
@@ -312,3 +312,6 @@
 
 # < /dev/urandom tr -dc _A-Z-a-z-0-9 | head -c${1:-80};echo;
 SERVICE_TOKEN = '2JMV2GQ2_vBoP8r5TZAvH6gAeQp_etxey19126QPdcyy9O9b73oyhQ-kEc7N1bUtlTy7FOHz2CeBLM6W'
+
+# Test runner with no database creation
+TEST_RUNNER = 'data_ocean.tests.NoDbTestRunner'
```
[the remark this answers is not in the document]
Comment (Iva-khar) [reply data_converter/settings_local.base.py:317]: done

File: data_ocean/transliteration/constants.py [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+    'Ч': 'Ch',
+    'Ш': 'Sh',
+    'Щ': 'Shch',
+    'Ъ': '',
+    'Ы': 'Y',
+    'Ь': '',
+    'Э': 'E',
+    'Ю': 'Iu',
+    'Я': 'Ia',
+    '’': '',
```
[the remark this answers is not in the document]
Comment (Iva-khar) [reply data_ocean/transliteration/constants.py:86]: done

[the remark this answers is not in the document]
Comment (Iva-khar) [reply data_ocean/transliteration/constants.py:86]: I can't delete '

File: data_ocean/tests.py (hunk 11 above)
Comment (Iva-khar) [reply data_ocean/tests.py:20]: done

**Review by roman-tiukh — APPROVED** [review]

**Review by PyDev777 — APPROVED** [review]

**Pull request merged**